astropy/stats/lombscargle/core.py

```diff
@@ -0,0 +1,338 @@
+
+    def _validate_frequency(self, frequency):
+        frequency = np.asanyarray(frequency)
+
+        if has_units(self.t):
+            frequency = units.Quantity(frequency)
+            try:
+                frequency = units.Quantity(frequency, unit=1./self.t.unit)
+            except units.UnitConversionError:
+                raise ValueError("Units of frequency not equivalent to "
+                                 "units of 1/t")
+        else:
+            assert not has_units(frequency)
+        return frequency
+
+    def _validate_t(self, t):
+        t = np.asanyarray(t)
+
+        if has_units(self.t):
+            t = units.Quantity(t)
+            try:
+                t = units.Quantity(t, unit=self.t.unit)
+            except units.UnitConversionError:
+                raise ValueError("Units of t not equivalent to "
+                                 "units of input self.t")
+        return t
+
+    def _power_unit(self, norm):
+        if has_units(self.y):
+            if self.dy is None and norm == 'psd':
+                return self.y.unit ** 2
+            else:
+                return units.dimensionless_unscaled
+        else:
+            return 1
+
+    def autofrequency(self, samples_per_peak=5, nyquist_factor=5,
+                      minimum_frequency=None, maximum_frequency=None):
+        """Determine a suitable frequency grid for data.
+
+        Note that this assumes the peak width is driven by the observational
+        baseline, which is generally a good assumption when the baseline is
+        much larger than the oscillation period.
+        If you are searching for periods longer than the baseline of your
+        observations, this may not perform well.
+
+        Even with a large baseline, be aware that the maximum frequency
+        returned is based on the concept of "average Nyquist frequency", which
+        may not be useful for irregularly-sampled data. The maximum frequency
+        can be adjusted via the nyquist_factor argument, or through the
+        maximum_frequency argument.
+
+        Parameters
+        ----------
+        samples_per_peak : float (optional, default=5)
+            The approximate number of desired samples across the typical peak
+        nyquist_factor : float (optional, default=5)
+            The multiple of the average nyquist frequency used to choose the
+            maximum frequency if maximum_frequency is not provided.
+        minimum_frequency : float (optional)
+            If specified, then use this minimum frequency rather than one
+            chosen based on the size of the baseline.
+        maximum_frequency : float (optional)
+            If specified, then use this maximum frequency rather than one
+            chosen based on the average nyquist frequency.
+
+        Returns
+        -------
+        frequency : ndarray or Quantity
+            The heuristically-determined optimal frequency bin
+        """
+        t = np.asanyarray(self.t)
```

Comment:
This can just be `t = self.t`, right? `broadcast_arrays` will have ensured this is an array.